agents/raritan_px3/fence_raritan_px3.py

```diff
@@ -0,0 +1,195 @@
+compatible with PX3 and seem to share the same BIOS, so this fence should
+work with PX2 as well).
+It has another command line prompt and totally other commands
+and output.
+
+It follows the concept of separating outlets and outletgroups (if created).
+You can reach outlets via a fixed, not changeble "plug" number
+(from 1-20 on my device).
+Additionally one can, but does not need to assign names to each plug.
+
+Plugs/outlets can be combined to outletgroups.
+There can be zero to N (N = No. outlets) outletgroups.
+
+While it's possible to create outletgroups with one plug, this does not
+make sense and might slow things down.
+
+--plug=X paramter can be:
+1. X == outlet No
+2. X == outlet Name (if one got assigned)
+3. X == outlet group Name
+
+-> One cannot reach a group by number
+-> Groups need an extra call (first single outlet devices are
+   searched for given No/Name, then OutletGroups
+"""
+
+
+class FenceRaritanPX3:
+	outlets={}
+	# Plug id of outlet
+	plug=None
+	outletgroups={}
+	# Group name if outlet plug id/name have not been found
+	group_name=None
+
+def px3_get_outlet_group(conn, options):
+
+	conn.send_eol("show outletgroups")
+	conn.expect(options["--command-prompt"], int(options["--shell-timeout"]))
+	for line in conn.after.splitlines():
+		split_line = line.split(" ")
+		"""
+		Groups always have a name assigned:
+		```
+		Outlet Group 1 - test:
+		Member outlets: 10-11
+		State:          2 on
+		```
+		"""
+		if len(split_line) == 5 and split_line[0] == "Outlet" and split_line[1] == "Group":
+			group_no = split_line[2]
+			group_name = split_line[4][:-1]
+
+		if len(split_line) > 0 and split_line[0] == "State:":
+			group_state = split_line[-1]
+			FenceRaritanPX3.outletgroups[group_no] = [ group_name, group_state ]
+	logging.debug("Outletgroups found:\n%s", FenceRaritanPX3.outletgroups)
+	return FenceRaritanPX3.outletgroups
+
+def px3_get_outlet_list(conn, options):
+
+	conn.send_eol("show outlets")
+	conn.expect(options["--command-prompt"], int(options["--shell-timeout"]))
+	for line in conn.after.splitlines():
+		split_line = line.split(" ")
+		"""
+		Plug with no name assigned:
+		```
+		Outlet 1:
+		Power state: On
+		```
+		"""
+		if len(split_line) == 2 and split_line[0] == "Outlet":
+			outlet_no = split_line[1][:-1]
+			outlet_name = ""
+		"""
+		Plug with name assigned:
+		```
+		Outlet 8 - Test:
+		Power state: On
+		```
+		"""
+		if len(split_line) == 4 and split_line[0] == "Outlet":
+			outlet_no = split_line[1]
+			outlet_name = split_line[3][:-1]
+
+		# fetch state of previously parsed outlet from next line/iter
+		if len(split_line) == 3 and split_line[0] == "Power" and split_line[1] == "state:":
+			outlet_state = split_line[2]
+			FenceRaritanPX3.outlets[outlet_no] = [outlet_name, outlet_state]
+	logging.debug("Outlets found:\n%s", FenceRaritanPX3.outlets)
+	return FenceRaritanPX3.outlets
+
+def get_power_status(conn, options):
+
+	if FenceRaritanPX3.plug:
```

Comment:
What status's does it report? on/off/unknown is what the fencing library expects.

Reply:
Yes, on/off

Originally On/Off as shown in the comment:
                Power state: On
(credentials removed):

I save all in an outlet dict, if necessary a second call also
storing outletgroup data is done:

 ./fence_raritan_px3 ... --plug=20  --action=list-status
1,,On
2,,On
3,,On
4,,On
5,,On
6,,On
7,,On
8,Testung,On
9,,On
10,,On
11,,On
12,,On
13,,On
14,mustang1,On
15,mustang2,On
16,mustang4,On
17,mustang3,On
18,,On
19,,On
20,,On

Some plugs/outlets do have names, some can only be accessed via ID.
Raw dict:
{'1': ['', 'On'], '2': ['', 'On'], '3': ['', 'On'], '4': ['', 'On'], '5': ['', 'On'], '6': ['', 'On'], '7': ['', 'On'], '8': ['Testung', 'On'],
'9': ['', 'On'], '10': ['', 'On'], '11': ['', 'On'], '12': ['', 'On'], '13': ['', 'On'], '14': ['mustang1', 'On'],
'15': ['mustang2', 'On'], '16': ['mustang4', 'On'], '17': ['mustang3', 'On'], '18': ['', 'On'], '19': ['', 'On'], '20': ['', 'On']}

I could (should?) already convert the uppercase to lower?
I do this later:
        if FenceRaritanPX3.plug:
                return FenceRaritanPX3.outlets[str(FenceRaritanPX3.plug)][1].lower()
        elif FenceRaritanPX3.group_name:
                return FenceRaritanPX3.outletgroups[FenceRaritanPX3.group_name][1].lower()

So on/off can be returned.

Reply:
Sounds good.

Yeah. That makes sense, to make it the same as the other agents, which usually maps the statuses.

Example:
https://github.com/ClusterLabs/fence-agents/blob/master/agents/vmware_rest/fence_vmware_rest.py#L14